backend/sources/facebook/connector/src/main/java/co/airy/core/sources/facebook/Connector.java

```diff
@@ -0,0 +1,218 @@
+                            }
+
+                            aggregate.setChannelId(message.getChannelId());
+
+                            return aggregate;
+                        })
+                .join(channelsTable, Conversation::getChannelId, (aggregate, channel) -> {
+                    aggregate.setChannel(channel);
+                    return aggregate;
+                });
+
+        // Send outbound messages
+        messageStream.filter((messageId, message) -> DeliveryState.PENDING.equals(message.getDeliveryState()))
+                .join(conversationTable, (message, conversation) -> new SendMessageRequest(conversation, message))
+                .mapValues(this::sendMessage)
+                .to(new ApplicationCommunicationMessages().name());
+
+        // Fetch missing metadata
+        conversationTable
+                // To avoid any redundant fetch contact operations the suppression interval should
+                // be higher than the timeout of the Facebook API
+                .suppress(Suppressed.untilTimeLimit(Duration.ofMillis(streams.getSuppressIntervalInMs()), Suppressed.BufferConfig.unbounded()))
+                .toStream()
+                .leftJoin(metadataTable, (conversation, metadataMap) -> {
+                    conversation.setMetadata(new HashMap<>(Optional.ofNullable(metadataMap).orElse(Map.of())));
+                    return conversation;
+                })
+                .filter(this::needsMetadataFetched)
+                .flatMap(this::fetchMetadata)
+                .to(new ApplicationCommunicationMetadata().name());
+
+
+        streams.start(builder.build(), appId);
+    }
+
+    private Message sendMessage(SendMessageRequest sendMessageRequest) {
+        final Message message = sendMessageRequest.getMessage();
+        final Conversation conversation = sendMessageRequest.getConversation();
+
+        try {
+            final String pageToken = conversation.getChannel().getToken();
+            final SendMessagePayload fbSendMessagePayload = mapper.fromSendMessageRequest(sendMessageRequest);
+
+            api.sendMessage(pageToken, fbSendMessagePayload);
+
+            updateDeliveryState(message, DeliveryState.DELIVERED);
+            return message;
+        } catch (ApiException e) {
+            log.error(String.format("Failed to send a message to Facebook \n SendMessageRequest: %s \n Error Message: %s \n", sendMessageRequest, e.getMessage()), e);
+        } catch (Exception e) {
+            log.error(String.format("Failed to send a message to Facebook \n SendMessageRequest: %s", sendMessageRequest), e);
+        }
+
+        updateDeliveryState(message, DeliveryState.FAILED);
+        return message;
+    }
+
+    private boolean needsMetadataFetched(String conversationId, Conversation conversation) {
+        final Map<String, String> metadata = conversation.getMetadata();
+        final String fetchState = metadata.get(MetadataKeys.Source.CONTACT_FETCH_STATE);
```

Comment:
at glance, I feel like `MetadataKeys.Source.Contact.FETCH_STATE` would be a better name/place for this?

Reply:
This was my initial idea, too, but the issue is that for the `conversations.list` endpoint we take everything in `source.contact.*` and map it to a response. Since I wanted to hide this value, I moved it up in the hierarchy.